ocaml/typing/typecore.ml

```diff
@@ -8664,6 +8632,43 @@ and type_expect_jane_syntax
   | Jexp_tuple x ->
       type_tuple
         ~loc ~env ~expected_mode ~ty_expected ~explanation ~attributes x
+  | Jexp_modes x ->
+      type_mode_expr
+        ~loc ~env ~expected_mode ~ty_expected ~explanation ~attributes x
+
+and type_mode_expr
+    ~loc ~env ~expected_mode ~ty_expected ~explanation ~attributes
+  : Jane_syntax.Modes.expression -> _ = function
+  | Coerce (modes, sbody) ->
+    let expected_mode = List.fold_left (fun expected_mode mode ->
+        let mode = (mode : Jane_syntax.Mode_expr.Const.t :> _ Location.loc) in
+        match mode.txt with
+        | "unique" ->
+          let expected_mode = mode_unique expected_mode in
+          expect_mode_cross env ty_expected expected_mode
+        | "once" ->
+          let expected_mode = expect_mode_cross env ty_expected expected_mode in
+          submode ~loc ~env ~reason:Other
+            (Value.min_with_linearity Linearity.once) expected_mode;
+          mode_once expected_mode
+        | "local" ->
+          let expected_mode = expect_mode_cross env ty_expected expected_mode in
+          submode ~loc ~env ~reason:Other
+            (Value.min_with_regionality Regionality.local) expected_mode;
+          mode_strictly_local expected_mode
+        | s ->
+          Misc.fatal_errorf "Unrecognized mode %s - should not parse" s
+        ) expected_mode modes.txt
+    in
+    let exp =
+      type_expect env expected_mode sbody (mk_expected ty_expected ?explanation)
+    in
+    {exp with
+     (* CR modes: We should consider not overriding [exp_loc] here -- that would
+        be more consistent to the typing of [Pexp_constraint].
+     *)
```

Comment:
That's good point - I wonder what's preventing us? If it's just the change of error messages, I think that's fine, and we should just do it (i.e., not override `exp_loc`)?

Reply:
This does change a lot of error messages. I don't think we should do it in this PR, one of whose benefits is that it's clearly a no-op to tests.